# Project 1: NumPy — Real Estate Market Analysis

**Role:** Data Analyst, Real Estate Company
**Goal:** Analyze property data using NumPy to identify patterns in property
prices and characteristics.

**Dataset:** [Housing Prices Regression Dataset](https://www.kaggle.com/datasets/denkuznetz/housing-prices-regression) (Kaggle)

**Concepts covered:** NumPy arrays, structured arrays, indexing, aggregation,
Boolean masking, broadcasting, sorting, and fancy indexing.


## 1. Load and Explore the Data
**Concepts: Structured Arrays, Indexing**

In [1]:
import numpy as np

# Load the CSV as a NumPy structured array.
# names=True reads the header row as field names.
# dtype=None lets NumPy infer each column's data type.
# encoding=None keeps text fields as regular Python strings.
data = np.genfromtxt("real_estate_dataset.csv", delimiter=",",
                      names=True, dtype=None, encoding=None)

print("Shape:", data.shape)
print("\nField names:", data.dtype.names)
print("\nField data types:")
for name in data.dtype.names:
    print(f"  {name}: {data.dtype[name]}")


Shape: (500,)

Field names: ('ID', 'Square_Feet', 'Num_Bedrooms', 'Num_Bathrooms', 'Num_Floors', 'Year_Built', 'Has_Garden', 'Has_Pool', 'Garage_Size', 'Location_Score', 'Distance_to_Center', 'Price')

Field data types:
  ID: int64
  Square_Feet: float64
  Num_Bedrooms: int64
  Num_Bathrooms: int64
  Num_Floors: int64
  Year_Built: int64
  Has_Garden: int64
  Has_Pool: int64
  Garage_Size: int64
  Location_Score: float64
  Distance_to_Center: float64
  Price: float64


In [2]:
# Display the first five records
print(data[:5])


[(1, 143.63502971, 1, 3, 3, 1967, 1, 1, 48, 8.2976312 ,  5.93573364, 602134.81674659)
 (2, 287.6785766 , 1, 2, 1, 1949, 0, 1, 37, 6.06146565, 10.8273922 , 591425.13538622)
 (3, 232.99848545, 1, 3, 2, 1923, 1, 0, 14, 2.91144248,  6.90459907, 464478.69687988)
 (4, 199.66462105, 5, 2, 2, 1918, 0, 0, 17, 2.07094918,  8.28401851, 583105.65599648)
 (5,  89.00466011, 4, 3, 3, 1999, 1, 0, 34, 1.52327786, 14.6482773 , 619879.14252279)]


**Interpretation:** The dataset loads as a 1-dimensional structured array,
where each element is one property (one row), and each field (column) can be
accessed by name, e.g. `data['Price']`. This is the NumPy equivalent of a
table, without needing pandas.

## 2. Analyze Property Prices
**Concepts: Indexing, Aggregation Functions**

In [3]:
prices = data['Price']

price_mean = np.mean(prices)
price_median = np.median(prices)
price_min = np.min(prices)
price_max = np.max(prices)
price_p25 = np.percentile(prices, 25)
price_p75 = np.percentile(prices, 75)

print(f"Mean price:   ${price_mean:,.2f}")
print(f"Median price: ${price_median:,.2f}")
print(f"Min price:    ${price_min:,.2f}")
print(f"Max price:    ${price_max:,.2f}")
print(f"25th percentile: ${price_p25:,.2f}")
print(f"75th percentile: ${price_p75:,.2f}")


Mean price:   $582,209.63
Median price: $574,724.11
Min price:    $276,892.47
Max price:    $960,678.27
25th percentile: $503,080.34
75th percentile: $665,942.30


**Interpretation:** The mean price ($582,210) is close to the
median ($574,724), suggesting prices are fairly evenly
distributed rather than skewed by a handful of extreme outliers. The
25th–75th percentile range shows where the "middle half" of the market
sits.

## 3. Analyze Property Size
**Concepts: Indexing, Aggregation Functions**

In [4]:
sizes = data['Square_Feet']

size_mean = np.mean(sizes)
size_median = np.median(sizes)
size_min = np.min(sizes)
size_max = np.max(sizes)

print(f"Mean size:   {size_mean:,.1f} sq ft")
print(f"Median size: {size_median:,.1f} sq ft")
print(f"Min size:    {size_min:,.1f} sq ft")
print(f"Max size:    {size_max:,.1f} sq ft")


Mean size:   174.6 sq ft
Median size: 178.3 sq ft
Min size:    51.3 sq ft
Max size:    298.2 sq ft


**Interpretation:** The gap between minimum and maximum size shows how
diverse the properties in this dataset are, from small starter homes to large
estates. A mean close to the median suggests sizes are fairly evenly
distributed, without extreme outliers dominating the average.

## 4. Calculate Price per Square Foot
**Concept: Element-wise Array Operations**

In [5]:
# Element-wise division: each property's price divided by its own size
price_per_sqft = data['Price'] / data['Square_Feet']

print(f"Mean price/sqft:   ${np.mean(price_per_sqft):,.2f}")
print(f"Median price/sqft: ${np.median(price_per_sqft):,.2f}")
print(f"Min price/sqft:    ${np.min(price_per_sqft):,.2f}")
print(f"Max price/sqft:    ${np.max(price_per_sqft):,.2f}")


Mean price/sqft:   $4,008.82
Median price/sqft: $3,361.19
Min price/sqft:    $1,636.48
Max price/sqft:    $11,572.99


**Interpretation:** Price per square foot standardizes properties of
different sizes so they can be compared fairly. A wide range between the
minimum and maximum suggests that factors other than size alone — like
location — are driving a large part of the price.

## 5. Identify Expensive Properties
**Concepts: Boolean Masking, Aggregation Functions**

In [6]:
# Boolean mask: True for properties priced in the top 25% (at or above the 75th percentile)
top25_mask = data['Price'] >= price_p75
expensive_properties = data[top25_mask]

n_expensive = len(expensive_properties)
avg_expensive_price = np.mean(expensive_properties['Price'])
overall_avg_price = np.mean(data['Price'])
difference = avg_expensive_price - overall_avg_price

print(f"Number of properties in top 25% by price: {n_expensive}")
print(f"Average price of top 25%:   ${avg_expensive_price:,.2f}")
print(f"Overall average price:      ${overall_avg_price:,.2f}")
print(f"Difference (top 25% vs overall): ${difference:,.2f}")


Number of properties in top 25% by price: 125
Average price of top 25%:   $744,142.53
Overall average price:      $582,209.63
Difference (top 25% vs overall): $161,932.90


**Interpretation:** The top 25% of properties average
$744,143, about $161,933 above the overall
market average of $582,210 — confirming a clear premium segment
exists well above typical market pricing.

## 6. Apply Multiple Conditions
**Concepts: Boolean Masking, Multiple Boolean Conditions**

In [7]:
avg_size = np.mean(data['Square_Feet'])

# Combine three Boolean conditions with & (all must be true)
combined_mask = (
    (data['Square_Feet'] > avg_size) &
    (data['Price'] >= price_p75) &
    (data['Location_Score'] >= 8)
)

premium_properties = data[combined_mask]

print(f"Properties meeting all 3 conditions: {len(premium_properties)}")
print("\nSample of matching properties:")
print(premium_properties[['Square_Feet', 'Price', 'Location_Score']][:5])


Properties meeting all 3 conditions: 17

Sample of matching properties:
[(266.54403644, 875352.54518835, 9.37378438)
 (257.18437729, 695024.31070497, 9.94981369)
 (282.42441309, 787131.56969077, 9.01383097)
 (254.50369148, 743412.14388911, 9.92442289)
 (277.06647149, 960678.27429119, 9.03757572)]


**Interpretation:** Only 17 properties (out of 500) satisfy
all three conditions at once — above-average size, top-quartile price, and a
location score of 8+. This is a small fraction of the market, indicating
that large size, high price, and prime location don't automatically come
together; properties combining all three represent a distinct, elite
segment.

## 7. Compare Prices to the Average
**Concept: Broadcasting**

In [8]:
# Broadcasting: subtracting a single scalar (overall_avg_price) from every
# element in the prices array, without writing a loop
price_diff_from_avg = data['Price'] - overall_avg_price

above_avg_mask = price_diff_from_avg > 0
below_avg_mask = price_diff_from_avg < 0

print(f"Properties above average: {np.sum(above_avg_mask)}")
print(f"Properties below average: {np.sum(below_avg_mask)}")
print(f"\nExample differences (first 5 properties): {price_diff_from_avg[:5]}")


Properties above average: 240
Properties below average: 260

Example differences (first 5 properties): [  19925.18721803    9215.50585766 -117730.93264868     896.02646792
   37669.51299423]


**How broadcasting was applied:** `overall_avg_price` is a single number
(a scalar), while `data['Price']` is an array of hundreds of values. NumPy
automatically "broadcasts" the single scalar across every element of the
array, subtracting it from each property's price in one operation, with no
explicit loop required.

## 8. Find the 10 Most Expensive Properties
**Concepts: Sorting, Fancy Indexing**

In [9]:
# argsort returns the indices that would sort the array; [::-1] reverses
# it to descending order, and [:10] takes the top 10 indices
top10_idx = np.argsort(data['Price'])[::-1][:10]

# Fancy indexing: passing an array of indices to pull out multiple
# specific rows/fields at once
top10 = data[top10_idx]

print("10 Most Expensive Properties:")
print(top10[['ID', 'Square_Feet', 'Num_Bedrooms', 'Location_Score', 'Price']])


10 Most Expensive Properties:
[(151, 277.06647149, 5, 9.03757572, 960678.27429119)
 ( 53, 284.87473539, 5, 7.60854345, 909199.56036649)
 (201, 210.50791154, 5, 8.18026749, 885909.21527569)
 (  8, 266.54403644, 5, 9.37378438, 875352.54518835)
 (474, 272.78577842, 5, 6.32216598, 867558.29676918)
 (500, 296.55268612, 4, 7.89432194, 864299.5002175 )
 (108, 238.88778464, 5, 2.53821078, 860013.25820393)
 (368, 276.33766049, 4, 3.85613513, 857487.18174448)
 (379, 253.61712065, 5, 7.42824975, 854374.5244548 )
 ( 70, 296.72173415, 4, 4.0667407 , 854304.12719063)]


**Interpretation:** Sorting by price and using fancy indexing lets us
pull out the full records of the 10 highest-priced properties at once,
including every field, rather than just the price values alone.

## 9. Final Analysis

In [10]:
top10_avg_size = np.mean(top10['Square_Feet'])
top10_avg_beds = np.mean(top10['Num_Bedrooms'])
top10_avg_location = np.mean(top10['Location_Score'])

overall_avg_size = np.mean(data['Square_Feet'])
overall_avg_beds = np.mean(data['Num_Bedrooms'])
overall_avg_location = np.mean(data['Location_Score'])

print(f"{'Metric':<20}{'Top 10':>12}{'Overall Avg':>15}")
print(f"{'Size (sqft)':<20}{top10_avg_size:>12,.0f}{overall_avg_size:>15,.0f}")
print(f"{'Bedrooms':<20}{top10_avg_beds:>12.1f}{overall_avg_beds:>15.1f}")
print(f"{'Location Score':<20}{top10_avg_location:>12.1f}{overall_avg_location:>15.1f}")


Metric                    Top 10    Overall Avg
Size (sqft)                  267            175
Bedrooms                     4.7            3.0
Location Score               6.6            5.2


**What characteristics tend to be associated with the most expensive
properties in this dataset?**

The 10 most expensive properties average 267 sq ft versus a
dataset-wide average of 175 sq ft, 4.7 bedrooms versus
3.0 overall, and a location score of 6.6 versus
5.2 overall. All three metrics are higher among the top 10,
suggesting that size, bedroom count, and location quality are jointly
associated with premium pricing, with location score showing one of the
largest relative gaps.

**Business Insight:** Property size and location score are the two factors
most consistently elevated among the highest-priced homes in this dataset.
For a real estate company, this means marketing and pricing strategy for
premium listings should emphasize square footage and neighborhood quality
above secondary features like pool or garden presence. Investors looking to
maximize resale value should prioritize acquiring larger properties in
highly-rated locations, since these two traits — more than any single
amenity — appear to command the steepest price premiums in this
market.